File: src/com/castlefrog/agl/domains/backgammon/BackgammonSimulator.java
package com.castlefrog.agl.domains.backgammon;

import com.castlefrog.agl.Adversarial2AgentSimulator;
import com.castlefrog.agl.IllegalActionException;
import com.castlefrog.agl.TurnType;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

/**
 * Backgammon has about 100 actions per game.
 */
public final class BackgammonSimulator extends Adversarial2AgentSimulator<BackgammonState, BackgammonAction> {
    private static final TurnType TURN_TYPE = TurnType.SEQUENTIAL;

    private BackgammonSimulator(BackgammonState state) {
        legalActions_ = new ArrayList<>();
        legalActions_.add(new ArrayList<BackgammonAction>());
        legalActions_.add(new ArrayList<BackgammonAction>());
        setState(state);
    }

    private BackgammonSimulator(BackgammonSimulator simulator) {
        super(simulator);
    }

    public BackgammonSimulator copy() {
        return new BackgammonSimulator(this);
    }

    public static BackgammonSimulator create(BackgammonState state) {
        return new BackgammonSimulator(state);
    }

    public void setState(BackgammonState state) {
        state_ = state;
        computeRewards();
        computeLegalActions();
    }

    public void stateTransition(List<BackgammonAction> actions) {
        BackgammonAction action = actions.get(state_.getAgentTurn());
        if (!legalActions_.get(state_.getAgentTurn()).contains(action)) {
            throw new IllegalActionException(action, state_);
        }

        byte[] locations = state_.getLocations();

        for (int i = 0; i < action.size(); i++) {
            int from = action.getMove(i).getFrom();
            int distance = action.getMove(i).getDistance();
            byte piece;

            if (locations[from] > 0) {
                piece = 1;
            } else {
                piece = -1;
            }
            int to = from + distance * piece;
            if (to > 0 && to < BackgammonState.getNumberOfLocations() - 1) {
                if (locations[to] * piece < 0) {
                    locations[to] = piece;
                    if (piece > 0) {
                        locations[25] -= piece;
                    } else {
                        locations[0] -= piece;
                    }
                } else {
                    locations[to] += piece;
                }
            }
            locations[from] -= piece;
        }
        byte[] dice = new byte[] {
                (byte) (Math.random() * BackgammonState.getNumberOfDieFaces() + 1),
                (byte) (Math.random() * BackgammonState.getNumberOfDieFaces() + 1) };
        state_ = new BackgammonState(locations, dice, (state_.getAgentTurn() + 1) % 2);
        computeRewards();
        computeLegalActions();
    }

    private void computeLegalActions() {
        clearLegalActions();
        List<BackgammonAction> legalActions = legalActions_.get(state_.getAgentTurn());
        byte[] locations = state_.getLocations();
        byte[] dice = state_.getDice();
        int piece;
        byte[] values;
        int depth;

        if (getRewards()[0] == 0) {
            if (state_.getAgentTurn() == 0) {
                piece = 1;
            } else {
                piece = -1;
            }

            if (dice[0] == dice[1]) {
                values = new byte[] {dice[0]};
            } else {
                values = dice;
            }

            if (dice[0] == dice[1]) {
                depth = 4;
            } else {
                depth = 2;
            }

            // Simplify the board
            for (int i = 0; i < BackgammonState.getNumberOfLocations(); i++) {
                if (locations[i] * piece == -1) {
                    locations[i] = 0;
                }
            }

            legalActions = dfs(locations, new LinkedList<BackgammonMove>(), values, piece, depth);

            // Prune moves that are too small
            int max = 0;
            for (BackgammonAction legalAction : legalActions) {
                if (legalAction.size() > max) {
                    max = legalAction.size();
                }
            }
            for (int i = 0; i < legalActions.size(); i++) {
                if (legalActions.get(i).size() != max) {
                    legalActions.remove(i--);
                }
            }
        }
    }

    private List<BackgammonAction> dfs(byte[] locations,
            LinkedList<BackgammonMove> moves, byte[] values, int piece,
            int depth) {
        List<BackgammonAction> legalActions = new ArrayList<>();
        int limit = BackgammonState.getNumberOfLocations();
        int start = 0;

        if (piece > 0 && locations[0] > 0) {
            limit = 1;
        } else if (piece < 0 && locations[25] < 0) {
            start = 25;
        }

        boolean moveOff = canMoveOff(locations, piece);
        for (int i = start; i < limit; i++) {
            if (locations[i] * piece >= 1) {
                for (int j = 0; j < values.length; j++) {
                    if (canMove(i, values[j], moveOff)) {
                        BackgammonMove move = BackgammonMove.valueOf(i,
                                values[j]);
                        if (moves.isEmpty() || move.compareTo(moves.getLast()) * piece >= 0) {
                            moves.addLast(move);
                            if (depth > 1) {
                                locations[i] -= piece;
                                int next = i + values[j] * piece;
                                if (next > 0 && next < BackgammonState.getNumberOfLocations() - 1) {
                                    locations[next] += piece;
                                }
                                int k = 0;
                                if (values.length == 2) {
                                    if (j == 0) {
                                        k = 1;
                                    } else {
                                        k = 0;
                                    }
                                }
                                legalActions.addAll(dfs(locations, moves, new byte[] {values[k]}, piece, depth - 1));
                                if (next > 0 && next < BackgammonState.getNumberOfLocations() - 1) {
                                    locations[next] -= piece;
                                }
                                locations[i] += piece;
                            } else {
                                legalActions.add(new BackgammonAction(moves));
                            }
                            moves.removeLast();
                        }
                    }
                }
            }
        }
        if (legalActions.size() == 0) {
            legalActions.add(new BackgammonAction(moves));
        }
        return legalActions;
    }

    private boolean canMove(int location, int distance, boolean moveOff) {
        if (state_.getAgentTurn() == 0) {
            int next = location + distance;
            return (next < BackgammonState.getNumberOfLocations() - 1 && state_.getLocation(next) >= -1) ||
                   (moveOff && next >= BackgammonState.getNumberOfLocations() - 1);
        } else {
            int next = location - distance;
            return (next > 0 && state_.getLocation(next) <= 1) || (moveOff && next <= 0);
        }
    }

    /**
     * Checks if a player can start moving pieces off of the board.
     * @return true if legal to move off board.
     */
    private boolean canMoveOff(byte[] locations, int piece) {
        if (piece > 0) {
            for (int i = 0; i < 19; i += 1) {
                if (locations[i] > 0) {
                    return false;
                }
            }
        } else {
            for (int i = 7; i < BackgammonState.getNumberOfLocations(); i += 1) {
                if (locations[i] < 0) {
                    return false;
                }
            }
        }
        return true;
    }

    public void computeRewards() {
        boolean pos = false, neg = false;
        for (int i = 0; i < BackgammonState.getNumberOfLocations(); i += 1) {
            if (state_.getLocation(i) > 0) {
                pos = true;
            } else if (state_.getLocation(i) < 0) {
                neg = true;
            }
        }
        if (!pos) {
            rewards_ = REWARDS_AGENT1_WINS;
        } else if (!neg) {
            rewards_ = REWARDS_AGENT2_WINS;
        } else {
            rewards_ = REWARDS_NEUTRAL;
        }
    }

    public TurnType getTurnType() {
        return TURN_TYPE;
    }
}

File: src/main/kotlin/com/castlefrog/agl/domains/backgammon/BackgammonSimulator.kt
package com.castlefrog.agl.domains.backgammon

import com.castlefrog.agl.Simulator
import com.castlefrog.agl.domains.AdversarialRewards
import com.castlefrog.agl.domains.nextPlayerTurnSequential
import kotlin.collections.ArrayList
import kotlin.math.max
import kotlin.random.Random

/**
 * Classic game of Backgammon
 */
class BackgammonSimulator(private val random: Random = Random) : Simulator<BackgammonState, BackgammonAction> {

    override val initialState: BackgammonState
        get() {
            val agentTurn: Int = random.nextInt(2)
            val dice: ByteArray = byteArrayOf(
                when (random.nextInt(15)) {
                    0 -> 1
                    1, 2 -> 2
                    3, 4, 5 -> 3
                    6, 7, 8, 9 -> 4
                    else -> 5
                },
                (random.nextInt(BackgammonState.N_DIE_FACES)).toByte()
            )
            return BackgammonState(dice = dice, agentTurn = agentTurn)
        }

    override fun calculateRewards(state: BackgammonState): IntArray {
        var pos = false
        var neg = false
        for (i in 0 until BackgammonState.N_LOCATIONS) {
            if (!pos && state.locations[i] > 0) {
                pos = true
            } else if (!neg && state.locations[i] < 0) {
                neg = true
            }
        }
        return if (!pos) {
            AdversarialRewards.BLACK_WINS
        } else if (!neg) {
            AdversarialRewards.WHITE_WINS
        } else {
            AdversarialRewards.NEUTRAL
        }
    }

    override fun calculateLegalActions(state: BackgammonState): List<List<BackgammonAction>> {
        val legalActions = arrayListOf<MutableList<BackgammonAction>>(ArrayList(), ArrayList())

        val rewards = calculateRewards(state)
        if (rewards[0] == 0) {
            val piece = if (state.agentTurn == 0) 1 else -1

            val dice = state.dice
            val values = if (dice[0] == dice[1]) {
                intArrayOf(dice[0] + 1)
            } else {
                intArrayOf(dice[0] + 1, dice[1] + 1)
            }

            val depth = if (dice[0] == dice[1]) 4 else 2

            val tempLegalActions = dfs(
                state.locations, ArrayList(),
                values, piece, depth, state.agentTurn
            )

            // only allow actions that are tied for using most moves
            var max = 0
            for ((moves) in tempLegalActions) {
                max = max(max, moves.size)
            }
            tempLegalActions
                .filter { it.moves.size == max }
                .forEach { legalActions[state.agentTurn].add(it) }
        }
        return legalActions
    }

    override fun stateTransition(state: BackgammonState, actions: Map<Int, BackgammonAction>): BackgammonState {
        val action = actions[state.agentTurn]
        val legalActions = calculateLegalActions(state)
        if (action === null || !legalActions[state.agentTurn].contains(action)) {
            throw IllegalArgumentException("Illegal action, $action, from state, $state")
        }

        val locations = state.locations

        for ((from, distance) in action.moves) {
            val piece: Byte = if (locations[from] > 0) 1 else -1
            val to = from + distance * piece
            if (to > 0 && to < BackgammonState.N_LOCATIONS - 1) {
                if (locations[to] * piece < 0) {
                    locations[to] = piece
                    if (piece > 0) {
                        locations[25] = (locations[25] - piece).toByte()
                    } else {
                        locations[0] = (locations[0] - piece).toByte()
                    }
                } else {
                    locations[to] = (locations[to] + piece).toByte()
                }
            }
            locations[from] = (locations[from] - piece).toByte()
        }
        val dice = byteArrayOf(
            random.nextInt(BackgammonState.N_DIE_FACES).toByte(),
            random.nextInt(BackgammonState.N_DIE_FACES).toByte()
        )
        return BackgammonState(locations, dice, nextPlayerTurnSequential(state.agentTurn, BackgammonState.N_PLAYERS))
    }

    companion object {

        private const val TURN_PLAYER_1 = 0

        private fun dfs(
            locations: ByteArray,
            moves: ArrayList<BackgammonMove>,
            values: IntArray,
            piece: Int,
            depth: Int,
            agentTurn: Int
        ): MutableList<BackgammonAction> {
            val legalActions = ArrayList<BackgammonAction>()
            val limit = if (piece > 0 && locations[0] > 0) 1 else BackgammonState.N_LOCATIONS
            val start = if (piece < 0 && locations[25] < 0) 25 else 0
            val moveOff = canMoveOff(locations, piece)

            for (i in start until limit) {
                if (locations[i] * piece >= 1) {
                    for (j in values.indices) {
                        if (canMove(i, values[j], moveOff, agentTurn, locations)) {
                            val move = BackgammonMove.valueOf(i, values[j])
                            if (moves.isEmpty() || move.compareTo(moves.last()) * piece >= 0) {
                                moves.add(move)
                                if (depth > 1) {
                                    locations[i] = (locations[i] - piece).toByte()
                                    val next = i + values[j] * piece
                                    if (next > 0 && next < BackgammonState.N_LOCATIONS - 1) {
                                        locations[next] = (locations[next] + piece).toByte()
                                    }
                                    var k = 0
                                    if (values.size == 2) {
                                        k = if (j == 0) 1 else 0
                                    }
                                    legalActions.addAll(
                                        dfs(
                                            locations, moves, intArrayOf(values[k]),
                                            piece, depth - 1, agentTurn
                                        )
                                    )
                                    if (next > 0 && next < BackgammonState.N_LOCATIONS - 1) {
                                        locations[next] = (locations[next] - piece).toByte()
                                    }
                                    locations[i] = (locations[i] + piece).toByte()
                                } else {
                                    legalActions.add(BackgammonAction(moves.toHashSet()))
                                }
                                moves.removeLast()
                            }
                        }
                    }
                }
            }
            if (legalActions.size == 0) {
                legalActions.add(BackgammonAction(moves.toHashSet()))
            }
            return legalActions
        }

        private fun canMove(
            location: Int,
            distance: Int,
            moveOff: Boolean,
            agentTurn: Int,
            locations: ByteArray
        ): Boolean {
            return if (agentTurn == TURN_PLAYER_1) {
                val next = location + distance
                next < BackgammonState.N_LOCATIONS - 1 && locations[next] >= -1 ||
                        moveOff && next >= BackgammonState.N_LOCATIONS - 1
            } else {
                val next = location - distance
                next > 0 && locations[next] <= 1 || moveOff && next <= 0
            }
        }

        /**
         * Checks if a player can start moving pieces off of the board.
         * @return true if legal to move off board.
         */
        private fun canMoveOff(locations: ByteArray, piece: Int): Boolean {
            if (piece > 0) {
                (0..18)
                    .filter { locations[it] > 0 }
                    .forEach { _ -> return false }
            } else {
                (7 until BackgammonState.N_LOCATIONS)
                    .filter { locations[it] < 0 }
                    .forEach { _ -> return false }
            }
            return true
        }
    }
}
